modules/linear_boltzmann_solvers/discrete_ordinates_solver/sweep/spds/aah_spds.cc

```diff
@@ -0,0 +1,190 @@
+// SPDX-FileCopyrightText: 2024 The OpenSn Authors <https://open-sn.github.io/opensn/>
+// SPDX-License-Identifier: MIT
+
+#include "modules/linear_boltzmann_solvers/discrete_ordinates_solver/sweep/spds/aah_spds.h"
+#include "framework/mesh/mesh_continuum/mesh_continuum.h"
+#include "framework/logging/log.h"
+#include "framework/utils/timer.h"
+#include "framework/runtime.h"
+#include "caliper/cali.h"
+#include <boost/graph/topological_sort.hpp>
+#include <algorithm>
+
+namespace opensn
+{
+
+AAH_SPDS::AAH_SPDS(int id, const Vector3& omega, const MeshContinuum& grid, bool allow_cycles)
+  : SPDS(omega, grid), id_(id), allow_cycles_(allow_cycles)
+{
+  CALI_CXX_MARK_SCOPE("AAH_SPDS::AAH_SPDS");
+
+  // Populate Cell Relationships
+  size_t num_loc_cells = grid.local_cells.size();
+  std::vector<std::set<std::pair<int, double>>> cell_successors(num_loc_cells);
+  std::set<int> location_successors;
+  std::set<int> location_dependencies;
+
+  PopulateCellRelationships(omega, location_dependencies, location_successors, cell_successors);
+
+  location_successors_.reserve(location_successors.size());
+  for (auto v : location_successors)
+    location_successors_.push_back(v);
+
+  location_dependencies_.reserve(location_dependencies.size());
+  for (auto v : location_dependencies)
+    location_dependencies_.push_back(v);
+
+  // Create local cell graph
+  Graph local_cell_graph(num_loc_cells);
+
+  for (auto c = 0; c < num_loc_cells; ++c)
+    for (auto& successor : cell_successors[c])
+      boost::add_edge(c, successor.first, successor.second, local_cell_graph);
+
+  // Remove cycles
+  if (allow_cycles)
+  {
+    auto edges_to_remove = RemoveCyclicDependencies(local_cell_graph);
+    for (auto& edge_to_remove : edges_to_remove)
+      local_sweep_fas_.emplace_back(edge_to_remove.first, edge_to_remove.second);
+  }
+
+  // Generate topological ordering
+  spls_.item_id.clear();
+  boost::topological_sort(local_cell_graph, std::back_inserter(spls_.item_id));
+  std::reverse(spls_.item_id.begin(), spls_.item_id.end());
+  if (spls_.item_id.empty())
+  {
+    throw std::logic_error("AAH_SPDS: Cyclic dependencies found in the local cell graph.\n"
+                           "Cycles need to be allowed by the calling application.");
+  }
+
+  // Generate location-to-location dependencies
+  global_dependencies_.resize(opensn::mpi_comm.size());
+  CommunicateLocationDependencies(location_dependencies_, global_dependencies_);
+}
+
+void
+AAH_SPDS::BuildGlobalSweepFAS()
```

Comment:
please decide if we should spell out FAS and maybe add some short description

Reply:
I added a description, but decided to stick with `BuildGlobalSweepFAS` to keep things concise.